### PR TITLE
FINERACT-1971: Accrual for interest bearing loan with charge accrual based on submitted On Date fix

### PR DESCRIPTION
**Pull request by ruchiD**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 1]
```diff
@@ -1389,6 +1389,11 @@ private void applyPeriodicAccruals(final Collection<LoanTransaction> accruals) {
         }
     }
 
+    private LocalDate getDateForRangeCalculationWhenChargeAccrualOnSubmittedDate(LoanTransaction loanTransaction) {
+        return loanTransaction.getLoanChargesPaid().size() == 0 ? loanTransaction.getTransactionDate()
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: `isEmpty` might be better for readability.

However I dont really like this method. There are 2 places you check whether accrual transaction date to be used or the charge due date. Would you mind merge the conditions into just one place? You need a date , the method should decide entirely what date to be retrieved and hence to be used.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 2]
```diff
@@ -1389,6 +1387,12 @@ private void applyPeriodicAccruals(final Collection<LoanTransaction> accruals) {
         }
     }
 
+    private LocalDate getDateForRangeCalculation(LoanTransaction loanTransaction, boolean isChargeAccrualBasedOnSubmittedOnDate) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:1390]: It looks better, however i was wondering what if the loan charge is not a specific due date charge? Would it be better to call `getEffectiveDueDate()` method which covers the proper due date for installment charge as well?

Comment (ruchiD) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:1390]: Yes, getEffectiveDueDate() covers charge due dates properly, I will update the PR.

**Review by alberto-art3ch — APPROVED** [review]
LGTM

**Pull request merged**